integrations/tensorflow/e2e/iree_e2e_product_test_suite.bzl

```diff
@@ -0,0 +1,229 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Macro for building e2e tests from a single source with multiple flags."""
+
+load("//bindings/python:build_defs.oss.bzl", "iree_py_test")
+
+def normalize_dictionary(dictionary):
+    """Wraps every value of dictionary in a list if it isn't one already."""
+    for key, value in dictionary.items():
+        if type(value) != type([]):
+            dictionary[key] = [value]
+    return dictionary
+
+def dictionary_product(dictionary):
+    """Returns a named cartesian product of dictionary's values."""
+
+    # Converts {'a': [1, 2], 'b': [3, 4]} into
+    # [{'a': 1, 'b': 3}, {'a': 1, 'b': 4}, {'a': 2, 'b': 3}, {'a': 2, 'b': 4}]
+    product = [[]]
+    for values in dictionary.values():
+        # Iteratively grow the elements of the product.
+        product = [element + [value] for element in product for value in values]
+    dicts = [{k: v for k, v in zip(dictionary, element)} for element in product]
+    return dicts
+
+def iree_e2e_product_test_suite(
+        name,
+        srcs,
+        main,
+        flags_to_values,
```

Comment:
Standard tests have an `args` attribute. Can we use that naming and style instead to reuse a familiar API rather than creating a new one?

Reply:
Not sure I follow. Bazel `py_test`s have an `args` attribute that we're using in the macro that takes a sequence of strings representing a flag and a single value for that flag. Not sure how we could use that to generate a test suite?

Reply:
I renamed the test suite to `iree_e2e_cartesian_product_test_suite` :)

I think there's a benefit to the macro not needing to know the flag names, and it also means that the build rule directly mirrors the names used in the test file itself, where as before we were mapping `backends` to `target_backends`, `models` to `model`, `datasets` to `data`, `optimizer` to `optimizer_name` and so on.

This will be really useful for cleaning up things internally too.